janusgraph-core/src/main/java/org/janusgraph/graphdb/database/serialize/attribute/DateSerializer.java

```diff
@@ -26,7 +26,8 @@
 public class DateSerializer implements OrderPreservingSerializer<Date> {
 
     private final LongSerializer ls = LongSerializer.INSTANCE;
-    private final StdDateFormat dateFormat = StdDateFormat.instance;
+    // StdDateFormat is not thread-safe
+    private static final ThreadLocal<StdDateFormat> dateFormat = ThreadLocal.withInitial(StdDateFormat::new);
```

Comment:
Didn't check the logic but wanted to say that ThreadLocal data is not released if the thread is killed / stopped.
Which means that we can see data leak here if we don't properly remove threadlocal data

Reply:
Hey @porunov. Good point. Glad you mentioned it. I'll open another PR to fix this.

Reply:
Hi is there any resource that explains how/why ThreadLocal data is not garbage collected for a dead thread?

Reply:
Quickly reading javadoc I'm now in doubt about my words. Will need to check more info later. Maybe I'm just wrong. I do remember memory leaks with ThreadLocal  but maybe it's only reproducible with cached thread pool. I will check that later

Reply:
Here is a reference (from Sonar) to leaks happening when re-using threads: https://rules.sonarsource.com/java/RSPEC-5164
If you believe there is a leak, and as already mentioned on this PR/issue, we can always switch to using the thread-safe Java 8+ DateTimeFormatter.

I've created a branch with a candidate change set: https://github.com/JanusGraph/janusgraph/compare/master...cdegroc:java-date-time-formatterhttps://github.com/JanusGraph/janusgraph/compare/master...cdegroc:java-date-time-formatter

Reply:
> I do remember memory leaks with ThreadLocal 

I think you are referring to the case where the transaction (or even janusgraph) is closed but the thread (likely part of an externally managed thread pool) is reused for other affairs. E.g. https://github.com/JanusGraph/janusgraph/issues/1341

Given this StdDateFormat shouldn't be used in the first place since it is intended for internal use only, I agree we should switch to alternative solutions. That being said, I haven't reviewed @cdegroc 's thread-safe solution yet.